### ISSUE TITLE
state of Context Parallel & Flex attention

### ISSUE THREAD

**Issue by rakkit** [pytorch/torchtitan#2417]
Hi, i am trying to catch up what's the state of Context Parallel for Flex attention.

Per https://github.com/pytorch/torchtitan/pull/2145, CP + Flex attention is supported now (llama 3 & llama4) ? and does it means "generally" Flex attention can work with CP? For example sliding window attention?

And is this rope_cache https://github.com/pytorch/torchtitan/issues/2077 blocked Flex attention CP for qwen & GPT-OSS and/or is there any another reasons?

And is there any reason (its werid) why CP & Flex attention is turn off for [deepseek](https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/deepseek_v3/infra/parallelize.py#L71)?

**Comment by tianyu-l** [comment]
It should work in general. Needs to add the code and validation. cc @fegin

**Comment by rakkit** [comment]
thx @tianyu-l in that case (need bit of refactor), i have a noob question: why dont we map llama3/llama4 to qwen3/gpt-oss rope style (pre-compute cos/sin),  in the sense 1) compiler seems work much smooth on cos/sin float tensor and 2) QK-Norm is more or less "standard"  today and compile + full graph can directly fused Norm + rope into one kernel.

even we dont do this, keep fused norm + rope in qwen3/gpt-oss is still a good idea

**Comment by tianyu-l** [comment]
Conventionally, llama3 / llama4 's released code uses complex rope instead of cos/sin (and it also uses a context length extension method different from yarn which seems to be coupled with complex code). Besides, llama3/4 doesn't have qk-norm.

I don't think there're fundamental reasons -- maybe we should replace everything with compiler-friendly code, or even hand-written kernel lol.

Feel free to propose changes, after my refactor https://github.com/pytorch/torchtitan/pull/2386

cc @shuhuayu

**Comment by fegin** [comment]
@rakkit In general CP + Flex should work for all models. I probably have not eliminated all the exceptions. But the functionality is there. The only caveat is that we have not verified if CP and EP have any composability issues, which is my next step before completely roll out CP + Flex for all models.

**Comment by shuhuayu** [comment]
I agree that there are no fundamental reasons to unify rope implementations to the cos/sin style. Using complex numbers can have benefits in neatness and flexibility with other RoPE variants, but working better with compile seems to me more important.

**Comment by rakkit** [comment]
thanks a lot @fegin,  if i understand correctly qwen3 is almost "ready" for CP,  (SDPA + CP & FLEX (block / block-causal) + CP)

and gpt-oss (two attention mask) seems is a special case need flex +  `headtail` balancer? the current `ptrr` seems does not support for `dict[str, BlockMask]`.   (gpt-oss also needs some refactor to take positions and apply rope correctly)

**Comment by fegin** [comment]
@rakkit Yes, that's a good point. We are actually considering to use `headtail` as the default and allow users to use `ptrr`. It seems that `headtail` provides more bit-wise parity guarantee.

**Comment by rakkit** [comment]
thanks a lot for clarify!

**Comment by codestar12** [comment]
I have a question about CP. I have a private mirror of torchtitan where we have pinned the pytorch version to 2.10. I remember previously using CP w/ flex attention, especially when using block_causal masks resulted in loss diverging. I assume the recent enablement means this issue has been fixed. My question is, is there a stable version of pytorch either released, or expected to release where this should work correctly? In my testing after upgrading to 2.10 and merging in latest upstream torchtitan changes I still noticed differences in loss when using cp compared to just dp or dp + tp.

**Comment by fegin** [comment]
CP will produce different loss due to the accumulation order when 1) doing ring attention or 2) load balancing is enabled, which q, k and v are shuffled. Is the training converging or you are seeing the loss curve very different?

**Comment by codestar12** [comment]
The runs do seem to be converging, but the loss curves are indeed very different. I'll see if I can screen shot some examples.

**Comment by fegin** [comment]
Can you provide a more specific detail on the setting? Are you using the PTRR load balancer?
